Support non-trivial `classes_` in `LogisticRegression`

jcrist opened the pull request:
Scikit-Learn's `LogisticRegression` is a bit unique in that it natively supports complex labels (e.g. raw strings/categories/non-monotonically increasing ints), rather than requiring that the labels are pre-encoded. It does this by internally using a `LabelEncoder` during `fit`, then converting the predicted labels back to the original label dtype in `predict`.

This PR adds support for this in `cuml`, improving compatibility with sklearn. This required:

- A small improvement to `LabelEncoder` to support all the input types that `cuml` natively supports.
- Addition of a `LabelEncoder` in `LogisticRegression.fit`
- Changing how we store `classes_`. Previously this used the descriptor functionality to support different container types. However, `CumlArray`/`cupy`/`numba` don't support non-numeric types, so we can't use that to store the classes anymore. We now always store `classes_` as a numpy array. I think this is fine - the size of classes is small - and also makes us a bit more compatible with sklearn since we can better ensure our dtypes match theirs.
- Changing `predict` to convert the numeric output back into the original classes. This was complicated since `CumlArray`/`cupy`/`numba` don't support non-numeric types, which means that most of our existing `output_type` machinery fails for these cases. I hacked something in that I think is sufficient, but it's definitely a hack.
- Addition of a new test to check things work properly across dtypes and container types.

This is an alternative to #6328. The fix here doesn't add any additional state, and I believe the test cases added here provide better coverage of the behavior we're trying to ensure.

jcrist commented on python/cuml/cuml/linear_model/logistic_regression.pyx:
```diff
@@ -405,14 +400,64 @@ class LogisticRegression(UniversalBase,
+        if is_numeric:
+            if (self.classes_ == np.arange(nclasses)).all():
+                # Fast path for common case of monotonically increasing numeric classes
+                out = indices.to_output("cupy", output_dtype=self.classes_.dtype)
```
Note that any code working prior to this change will take this path.

csadorf reviewed (commented):
From my initial review, this looks overall very sound. It's a bit of hack to work around the CumlArray limitation of course, but that was expected and we can of course address that moving forward.

I have not had the chance yet to investigate whether our tests fully cover cases where we call `LinearRegression.predict()` internally and whether we maintain behavior in that case. Since we are replicating the api decorator in `predict()` there is a small chance that we are not covering all edge cases. That is not the only reason I'm not approving just yet.

jcrist commented:
Looks like there's some test failures that I missed fixing locally. Most look pretty straightforward - if we want to still try and get this in pre-patch I can work on resolving these tomorrow.

viclafargue reviewed (commented):
LGTM, would just add a small test to prevent any regression for issue https://github.com/rapidsai/cuml-accel/issues/94

jcrist commented:
That case should already sufficiently tested by `test_logistic_regression_complex_classes` added here (the `int` and `float` cases to be specific)

csadorf commented:
> Looks like there's some test failures that I missed fixing locally. Most look pretty straightforward - if we want to still try and get this in pre-patch I can work on resolving these tomorrow.

@jcrist and I just had a brief offline chat and agreed that it's worth trying to address the issues before code freeze.

jcrist commented:
I _think_ I've fixed all the test failures, but who knows. Also added a fix for categorical `y` support, which was another sklearn compatibility bug we had.

csadorf commented:
/merge
